File: exercises/03-sagas/before/Divergent.Frontend/app/modules/finance/ordersLoadedSubscriber.js
(function () {

    angular.module('app.services')
        .config(['backendCompositionServiceProvider',
            function (backendCompositionServiceProvider) {

                var requestId = 'orders-list';
                backendCompositionServiceProvider.registerViewModelSubscriberFactory(requestId,
                    ['$log', '$http', 'finance.config', function ($log, $http, config) {

                        var subscriber = function (viewModel) {
                            viewModel.subscribe('orders/loaded', function (evt, ctrlViewModel, args) {

                                var orderIds = args.ordersViewModelDictionary.keys;

                                var uri = config.apiUrl + '/prices/orders/total?orderIds=' + orderIds;
                                return $http.get(uri)
                                    .then(function (response) {

                                        angular.forEach(response.data, function (value, key) {
                                            args.ordersViewModelDictionary[key].orderTotalPrice = value;
                                        });
                                    });
                            });
                        };

                        return subscriber;
                    }]);

            }]);
}())

File: exercises/01-composite-ui/after/Divergent.Finance.ViewModelComposition/OrdersLoadedSubscriber.cs
using System.Net.Http;
using Divergent.Sales.ViewModelComposition.Events;
using ITOps.Json;
using Microsoft.AspNetCore.Mvc;
using ServiceComposer.AspNetCore;

namespace Divergent.Finance.ViewModelComposition;

public class OrdersLoadedSubscriber : ICompositionEventsSubscriber
{
    [HttpGet("/orders")]
    public void Subscribe(ICompositionEventsPublisher publisher)
    {
        publisher.Subscribe<OrdersLoaded>(async (ordersLoaded, httpRequest) =>
        {
            var orderIds = string.Join(",", ordersLoaded.OrderViewModelDictionary.Keys);

            // Hardcoded to simplify the exercise. In a production app, a config object could be injected.
            var url = $"http://localhost:20187/api/prices/orders/total?orderIds={orderIds}";
            var response = await new HttpClient().GetAsync(url);

            dynamic[] prices = await response.Content.AsExpandoArrayAsync();

            foreach (dynamic price in prices)
            {
                ordersLoaded.OrderViewModelDictionary[price.OrderId].OrderTotalPrice = price.Amount;
            }
        });
    }
}